src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,119 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_MAX_SIZE = 10;
+    private static final int MIN_SIZE = 0;
+    private T[] data;
+    private int size;
+
+    @SuppressWarnings("unchecked")
+    public ArrayList() {
+        data = (T[]) new Object[DEFAULT_MAX_SIZE];
+        size = MIN_SIZE;
+    }
+
     @Override
     public void add(T value) {
-
+        growIfFull();
+        data[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkAddRange(index);
+        growIfFull();
+        System.arraycopy(data, index, data, index + 1, size - index);
+        data[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int addedArraySize = list.size();
+        while (addedArraySize > data.length - size) {
+            grow();
+        }
+        for (int i = 0; i < addedArraySize; i++) {
+            data[size + i] = list.get(i);
+        }
+        size += addedArraySize;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkRange(index);
+        return data[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkRange(index);
+        data[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkRange(index);
+        T value = data[index];
+        if (size - 1 > index) {
+            System.arraycopy(data, index + 1, data, index, size - 1);
```

Comment:
```suggestion
            System.arraycopy(data, index + 1, data, index, size - index - 1);
```

Reply:
I do not agree with this change. This parameter indicates the size of the new array. And it has decreased as a result by only one